### PR TITLE
Implement metrics for infraattributesprocessor

### PR DESCRIPTION
**Pull request by liustanley**
### What does this PR do?

Implement metrics processor for infraattributesprocessor. Gets entities from resource attributes and applies entity tags and global tags as resource attributes.

### CODE REVIEW

File: comp/otelcol/otlp/components/processor/infraattributesprocessor/config.go [hunk 1]
```diff
@@ -7,13 +7,17 @@ package infraattributesprocessor
...
+	"github.com/DataDog/datadog-agent/comp/core/tagger/types"
 )
 
 // Config defines configuration for processor.
 type Config struct {
 	Metrics MetricInfraAttributes `mapstructure:"metrics"`
 	Logs    LogInfraAttributes    `mapstructure:"logs"`
 	Traces  TraceInfraAttributes  `mapstructure:"traces"`
+
+	Cardinality types.TagCardinality `mapstructure:"cardinality"`
```
Comment (dinooliva) [inline comp/otelcol/otlp/components/processor/infraattributesprocessor/config.go:20]: My understanding is that cardinality is per-tag/attribute so I don't see how this config entry will specify cardinality for everything.

Comment (dinooliva) [reply comp/otelcol/otlp/components/processor/infraattributesprocessor/config.go:20]: After further discussions with @dineshg13, `Cardinality` is indeed a property of the agent/collector so this is correct. Resolving.

File: comp/otelcol/otlp/components/processor/infraattributesprocessor/metrics.go [hunk 2]
```diff
@@ -7,24 +7,70 @@ package infraattributesprocessor
...
-		logger: set.Logger,
+func newInfraAttributesMetricProcessor(set processor.CreateSettings, cfg *Config, tagger tagger.Component) (*infraAttributesMetricProcessor, error) {
+	iamp := &infraAttributesMetricProcessor{
+		logger:      set.Logger,
+		tagger:      tagger,
+		cardinality: cfg.Cardinality,
 	}
 	set.Logger.Info("Metric Tag Enrichment configured")
-	return tesp, nil
+	return iamp, nil
```
Comment (dinooliva) [inline comp/otelcol/otlp/components/processor/infraattributesprocessor/metrics.go:34]: Nit - update logger message.

File: comp/otelcol/otlp/components/processor/infraattributesprocessor/factory.go [hunk 3]
```diff
@@ -16,28 +19,38 @@ import (
...
 
-func createMetricsProcessor(
+func (f *factory) createMetricsProcessor(
 	ctx context.Context,
 	set processor.CreateSettings,
 	cfg component.Config,
 	nextConsumer consumer.Metrics,
 ) (processor.Metrics, error) {
-	tep, err := newInfraAttributesMetricProcessor(set, cfg.(*Config))
+	tep, err := newInfraAttributesMetricProcessor(set, cfg.(*Config), f.tagger)
```
Comment (dinooliva) [inline comp/otelcol/otlp/components/processor/infraattributesprocessor/factory.go:53]: Nit, here and elsewhere - rename `tep` to `iap`

**Review by dinooliva — APPROVED** [review]

**Review by dineshg13 — APPROVED** [review]

File: comp/otelcol/otlp/collector.go [hunk 4]
```diff
@@ -114,10 +114,11 @@ func getComponents(s serializer.MetricSerializer, logsAgentChannel chan *message
 		errs = append(errs, err)
 	}
 
-	processors, err := processor.MakeFactoryMap(
-		batchprocessor.NewFactory(),
-		infraattributesprocessor.NewFactory(),
```
Comment (purple4reina) [inline comp/otelcol/otlp/collector.go:119]: Looks like for serverless, we'll now no longer have this `infraattributesprocessor`. What are we losing when that processor goes away?

Comment (liustanley) [reply comp/otelcol/otlp/collector.go:119]: The `infraattributesprocessor` was only [recently added](https://github.com/DataDog/datadog-agent/pull/25039) as a no-op processor, this is the first implementation of it so no functionality will be lost for serverless. This is only intended for use with the new otel-agent binary for now

Comment (purple4reina) [reply comp/otelcol/otlp/collector.go:119]: Okay thanks, that makes sense! Just so I understand, what will this processor do once it's completely written? Is it something that we'll want to support in AWS Lambda? The last question is as much for you as it is for me.

If we don't want it in serverless, then I want to make sure that we don't add extra code to our binary just for it to all be ignored during runtime. It's really important for us to keep the size of our binary down. So, as you work on this, you might want to consider using build tags to compile it out of the serverless binary all together. 

Let me know if you'll need a hand with any of this, happy to help.

Comment (purple4reina) [reply comp/otelcol/otlp/collector.go:119]: For example, this PR increases the size of our binary by about 16KB but adds no benefit. That's a miniscule amount, so I'm not real worried about it. But, it's something we'll want to keep an eye on.

**Review by purple4reina — APPROVED** [review]

**Review by purple4reina — APPROVED** [review]

**Review by dustmop — APPROVED** [review]
LGTM for agent-shared-components owned files

**Pull request merged**